src/server/client.rs

```diff
@@ -189,66 +202,77 @@ async fn handle_job_cancel(
-                        )
-                    }
-                }
-                job.state = JobState::Canceled
-            }
-        };
-    }
+        let job = state.get_job_mut(job_id).unwrap();
+        let canceled_ids : Vec<_> = canceled_tasks.iter().map(|tako_id| job.set_cancel_state(*tako_id)).collect();
+        let already_finished = job.n_tasks() - canceled_ids.len() as JobTaskCount;
+        (canceled_ids, already_finished)
+    };
+    ToClientMessage::CancelJobResponse(CancelJobResponse::Canceled(canceled_ids, already_finished))
+}
+
+fn make_program_def_for_task(program_def: &ProgramDefinition, job_id: JobId, task_id: JobTaskId) -> ProgramDefinition {
+    let job_id_str = job_id.to_string();
+    let task_id_str = task_id.to_string();
+
+    let make_replacement = |s: &str| s.replace("%{JOB_ID}", &job_id_str).replace("%{TASK_ID}", &task_id_str);
 
-    ToClientMessage::CancelJobResponse(CancelJobResponse::Canceled)
+    let mut def = program_def.clone();
+    def.env.insert("HQ_JOB_ID".to_string(), job_id.to_string());
+    def.env.insert("HQ_TASK_ID".to_string(), task_id.to_string());
+    def.stdout = def.stdout.and_then(|p| p.to_str().map(make_replacement).map(PathBuf::from));
+    def.stderr = def.stderr.and_then(|p| p.to_str().map(make_replacement).map(PathBuf::from));
+    def
 }
 
 async fn handle_submit(
     state_ref: &StateRef,
     tako_ref: &TakoServer,
     message: SubmitRequest,
 ) -> ToClientMessage {
-    let (task_def, task_id, program_def) = {
+    let make_task = |job_id, task_id, tako_id| {
+        let program_def = make_program_def_for_task(&message.spec, job_id, task_id);
+        let body = rmp_serde::to_vec_named(&program_def).unwrap();
+        TaskDef {
+            id: tako_id,
+            type_id: 0,
+            body,
+            keep: false,
+            observe: true,
+            n_outputs: 0,
+        }
+    };
+    let (task_defs, job_detail) = {
         let mut state = state_ref.get_mut();
-        let task_id = state.new_job_id();
-
-        let mut program_def = message.spec;
-
-        let stdout = format!("stdout.{}", task_id);
-        let stderr = format!("stderr.{}", task_id);
-        program_def.stdout = Some(message.cwd.join(stdout));
-        program_def.stderr = Some(message.cwd.join(stderr));
+        let job_id = state.new_job_id();
+        let task_count = match &message.job_type {
+            JobType::Simple => 1,
+            JobType::Array(a) => a.task_count(),
+        };
+        let tako_base_id = state.new_task_id(task_count);
+        let mut tako_task_id = tako_base_id;
 
-        let body = rmp_serde::to_vec_named(&program_def).unwrap();
-        let job = Job::new(task_id, message.name.clone(), program_def.clone());
+        let task_defs = match &message.job_type {
+            JobType::Simple => vec![make_task(job_id, 0, tako_task_id)],
+            JobType::Array(a) => a.iter().map(|tt_id| {
```

Comment:
```rust
a.iter().zip(0..a.task_count()).map(|(tt_id, index)|
```
but that's probably less readable.

Reply:
You are right, just it has to be modified to different start, as it is not counted from 0, but from tako_base_id

Btw: There is also .enumerate() so you suggestion can be also even more simplified.

Reply:
Nice, I forgot about that.